renderer: add show password functionality

salfel opened the pull request:
This PR aims to provide password showing functionality which is being requested by at least a dozen people as shown in https://github.com/hyprwm/hyprlock/issues/478. 

Users can now toggle the password visibility by pressing tab. I have also added trimming functionality, for when the password size exceeds the available space of the input field.

I have also added the `password_center` and `password_size` configuration options to control the size and alignment of the password.

salfel commented:
> Thanks for the contribution @salfel :) I will test this later. From a first look though, I don't like the additional `password_size` and `password_center` options, since `dots_size` and `dots_center` seem to configure the same thing. The password input field already has so many options and possible configurations.

Okay. A possible solution is to just leave the password_size and password_center and remove the dots_center and dots_size options, making the password_* options apply for both dots and the password in plain text

PaideiaDilemma commented:
I have some opinions on the general conception of the feature.

- We should never allow password rendering by default.
  So when hyprlock starts, it never shows the password without an explicit toggle. For now, just make on-click trigger the toggle. (there will be no `on-click` config option for the input field, so we can use it internally.)
- I suggest an option in the config that allows toggling the password visibility. Disabled by default.
- For now just reuse `dots_size` and `dots_center`. `dots_*` isn't really a good name for those options since we added 
  `dots_text_format` anyways. At some point I would like to migrate to something like this:
  ```
  input-field {
    allow_toggle_password_visibility = 1
    visual-feedback {
      style="dots" # either "dots", "hidden" or a char
      center=true
      scale=0.3
      spacing=0.4
      rounding=2
     }
  }
  ```

PaideiaDilemma commented:
> Okay, so a possible implementation of this would be to either render a eye, which when clicked, toggles the visibility. Another option would be to toggle the visibility when just clicking on the input field, rendering no eye icon.

Another idea would be to make the "eye icon" a separate widget altogether. So something like a "password-visibility-toggle" widget. That would be easier to implement and give the user maximum flexibility, because they can put it wherever they want to.

salfel commented:
> I have some opinions on the general conception of the feature.
> 
>     * We should never allow password rendering by default.
>       So when hyprlock starts, it never shows the password without an explicit toggle. For now, just make on-click trigger the toggle. (there will be no `on-click` config option for the input field, so we can use it internally.)
> 
>     * I suggest an option in the config that allows toggling the password visibility. Disabled by default.
> 
>     * For now just reuse `dots_size` and `dots_center`. `dots_*` isn't really a good name for those options since we added
>       `dots_text_format` anyways. At some point I would like to migrate to something like this:
>       ```
>       input-field {
>         allow_toggle_password_visibility = 1
>         visual-feedback {
>           style="dots" # either "dots", "hidden" or a char
>           center=true
>           scale=0.3
>           spacing=0.4
>           rounding=2
>          }
>       }
>       ```

Good idea. I will try to work on this the coming weeks, given that I'm soon having my final exams for highschool.

salfel commented:
> I have some opinions on the general conception of the feature.
> 
>     * We should never allow password rendering by default.
>       So when hyprlock starts, it never shows the password without an explicit toggle. For now, just make on-click trigger the toggle. (there will be no `on-click` config option for the input field, so we can use it internally.)
> 
>     * I suggest an option in the config that allows toggling the password visibility. Disabled by default.
> 
>     * For now just reuse `dots_size` and `dots_center`. `dots_*` isn't really a good name for those options since we added
>       `dots_text_format` anyways. At some point I would like to migrate to something like this:
>       ```
>       input-field {
>         allow_toggle_password_visibility = 1
>         visual-feedback {
>           style="dots" # either "dots", "hidden" or a char
>           center=true
>           scale=0.3
>           spacing=0.4
>           rounding=2
>          }
>       }
>       ```

Just noticed that this style of config wouldn't be possible without a change to the hyprlock configmanager, as he doesn't support nested categories to my understanding. In my opinion, the visual-feedback can be omitted here

PaideiaDilemma commented:
> Just noticed that this style of config wouldn't be possible without a change to the hyprlock configmanager, as he doesn't support nested categories to my understanding. In my opinion, the visual-feedback can be omitted here

Hyprlang does support nested categories and it should also work with anonymous keys. You don't have to worry about that in this PR though. Like I said for now I would just reuse the `dots_*` options until we make a bit of an overhaul to the input-field configuration.

salfel commented:
> > Another idea would be to make the "eye icon" a separate widget altogether. So something like a "password-visibility-toggle" widget. That would be easier to implement and give the user maximum flexibility, because they can put it wherever they want to.
> 
> That may not be such a good idea in my opinion. In my opinion, it would be better to just give the user a option, to whether the eye should be right or left from the input, if he wants to use it. Any other option would make it cumbersome for the user to place the eye correctly. I'd at least make this the default. Maybe with another option the user could move the eye around as they want to.
> 
> This however would mean, that the eye would have to be a part of the input field itself which in my opinion is not a big deal

@PaideiaDilemma

salfel commented on src/renderer/widgets/PasswordInputField.cpp:
```diff
@@ -195,25 +197,41 @@ void CPasswordInputField::updatePassword() {
+    double                                  offset;
+    CBox                                    inputFieldBox = {pos, size->value()};
+
+    while (true) {
```
Because it is not possible to know how much the length of the input changes, because that depends on the font, we need to compare its length with the maximum available length. In the while loop it then trims the password on every iteration, until it is smaller than the available area in the textInputBox. To make this faster, i have also abstracted away the getting of the pango layout, to get it's size much faster. This computation however, could also be cached as it still takes quite some time.

PaideiaDilemma commented:
HI @salfel 

I tested this today and it seems like the `on-click` handler is not working properly on my multi-monitor setup. Also I think you should move the logic for differentiating eye icon coordinates from the rest of the input-field into a seperate function, so we can reuse to code to also update the cursor shape in on hover.

Then regarding the while loop:
I don't think we want to trim the asset like you do now.
We want to render the asset once and just render what fits into the password field. That is already done for the placeholdertext.

```cpp
            // Cut the texture to the width of the input field
            glEnable(GL_SCISSOR_TEST);
            glScissor(inputFieldBox.x, inputFieldBox.y, inputFieldBox.w, inputFieldBox.h);
            g_pRenderer->renderTexture(ASSETBOX, currAsset->texture, data.opacity * fade.a->value(), 0);
            glScissor(0, 0, viewport.x, viewport.y);
            glDisable(GL_SCISSOR_TEST);

```

You can check that out in action when the with of the input field resizes.
For example with a config like this.
```
nput-field {
  monitor=
  size=50, 50
  capslock_color=rgb(CB7459)
  dots_rounding=0
  dots_size=0.35
  dots_spacing=0.1
  dots_text_format=*
  fade_on_empty=false
  font_color=rgb(8F8F8F)
  font_family=Noto Sans
  inner_color=rgba(00000000)
  outer_color=rgba(33ccff88) rgba(00ff9988) 45deg
  check_color=rgba(33ccffee) rgba(00ff99ee) 405deg
  fail_color=rgba(ff6633ee) rgba(ff0066ee) 45deg
  outline_thickness=4
  placeholder_text=<span font_family="Noto Sans">$PROMPT</span>
  fail_text=<span font_family="Noto Sans">$FAIL <b>($ATTEMPTS)</b></span>
  position=0, -5%
  rounding=-1
  halign=center
  valign=center
}
```

I like how you embedded the eye icons!

salfel commented:
@PaideiaDilemma 
Thanks for the recommendations. Will try to fix the on-click. Regarding the on-hover, this will require the addition of a onCursorMove function, as the onHover function is only triggered when the pointer enters the widget. Will evaluate what the best solution for this would be

salfel commented:
i'd consider this feature now done, let me know if you find something to improve

salfel commented:
Sorry about that, i just thought that a dots_* prefix might be more explicit, considering that it does only affect the dots and not the text password. This cound however just be pointed out in the docs. Should i then remove the dots_* prefix?

salfel commented:
Hey @PaideiaDilemma, sorry to bother you. Could you please take a look at this, it seems that you have forgotten this

PaideiaDilemma commented:
Didn't forget, but i am occupied with fixing stuff atm. You didn't ping me either.

I will review it soon

PaideiaDilemma reviewed (changes requested):
Sorry that it took a bit to review...

I typed an entire review, exited Hyprland on accident and by browser failed to save it. Typing it again... (I am finally removed my keybinding for exiting hyprland.)

This is definitely not ready though. The password input field has so much logic. It's not easy to get it right.

Bugs:
- `input-field:dots_text_format` is broken
- After hitting enter on my password, the eye disappears, but the dots still show, which doesn't look good.
- Cursor still changes when the eye is not even visible.

Things:
- Don't make the password show/hide boolean global. Each input-field show have it's own state for that. When you click show on one monitor, it shouldn't show on another monitor. Only where I clicked.
- still lots of breaking changes.
- use `std::hash` for the password resourceIds. return a reference in `getPasswordBuffer`. Don't copy the password if it's not needed.
- I personally think pressing show and typing should reset the show state. That is because we need to render the whole password buffer each time and can't make a typing animation like we have with dots. Rendering each character separately would have the nice effect that it could use the exact animations and stuff, but I don't think the asyncResourceGatherer allows us to do that in an acceptable way currently.

I have an opinionated commit here that addresses some stuff if you are interested: https://github.com/PaideiaDilemma/hyprlock/commit/d55a816f915e2c4f211f5489ba1f41cfcc6c354d

salfel commented:
> Sorry that it took a bit to review...
> 
> I typed an entire review, exited Hyprland on accident and by browser failed to save it. Typing it again... (I am finally removed my keybinding for exiting hyprland.)
> 
> This is definitely not ready though. The password input field has so much logic. It's not easy to get it right.
> 
> Bugs:
> - `input-field:dots_text_format` is broken
> - After hitting enter on my password, the eye disappears, but the dots still show, which doesn't look good.
> - Cursor still changes when the eye is not even visible.
> 
> Things:
> - Don't make the password show/hide boolean global. Each input-field show have it's own state for that. When you click show on one monitor, it shouldn't show on another monitor. Only where I clicked.
> - still lots of breaking changes.
> - use `std::hash` for the password resourceIds. return a reference in `getPasswordBuffer`. Don't copy the password if it's not needed.
> - I personally think pressing show and typing should reset the show state. That is because we need to render the whole password buffer each time and can't make a typing animation like we have with dots. Rendering each character separately would have the nice effect that it could use the exact animations and stuff, but I don't think the asyncResourceGatherer allows us to do that in an acceptable way currently.
> 
> I have an opinionated commit here that addresses some stuff if you are interested: https://github.com/PaideiaDilemma/hyprlock/commit/d55a816f915e2c4f211f5489ba1f41cfcc6c354d
> 

Will take a look once I get home, thanks for the review and suggestions, sorry for being a bit sloppy with the pr, I'm still kinda new to this

salfel commented:
> `input-field:dots_text_format` is broken

i have named it dots_format

salfel commented:
i have now moved the show variable from hyprlock to each passwordinput widget. 

Currently the behaviour is as follows, when clicking tab, it just toggles everything regardless of its vaiue

Some proposals i have:
- when pressing tab, only toggle the visibility of the current focused window
- when pressing tab, if all passwords are hidden, show all of them, if at least one is shown, hide all

We could also remove tab for changing the visibility, though that would mean that users would have to use the mouse to toggle the  password visibility, which isn't that nice, as the hypr ecosystem tries to be very keyboard focused

salfel commented:
regarding the breaking changes, we could maybe leave the old config options and mark them as deprecated, and use a the new config. 

Feel free to make some suggestions about naming etc.

@PaideiaDilemma

PaideiaDilemma reviewed (changes requested):
@salfel 

> Feel free to make some suggestions about naming etc.

I think I have not managed to communicate my point about that well. Sorry for that.
I highly suggest you not change any options in this pr (we can try to clean things up later on). Adding ok. Changing means BREAKING . Exiting configurations will break and stop working the way they did before. Try to avoid that.
And if you do, because it just makes sense, please add a clear description about the option changes you plan to do. I don't want to have to look into ConfigManager.cpp to check.

> Cursor still changes when the eye is not even visible.

This still exists.

Other than that, looks pretty good tbh.

salfel commented:
> Cursor still changes when the eye is not even visible.

I have checked this again, and it seems to be working fine with any combination of toggle_password_visibility and hide_eye.

salfel commented:
@PaideiaDilemma 
sorry, i didn't understand you first. I hope this fixed it now. I also added some helper to not call wl on every mouse movement. Hope you are also happy with the backward comp change i added

salfel commented:
@PaideiaDilemma

vaxerski reviewed (commented):
some style stuff in the meantime.

salfel commented:
sorry it took so long to follow up no this pr. I was rather busy.

salfel commented:
I should have enough time right now. 

Regarding the keybind, that seems fair. Maybe we could also add a config option to set the keybind for this along with other actions. Just a thought for later on

About the refactor, ill look into it and see what would make sense

> I feel like tab is too easy to hit on accident.
> 
> How about Ctrl+h?
> 
> I also think `CPasswordInputField::draw` is getting a little out of hand generally so I probably would refactor that a bit. But that can be done later as well. The logic seems ok.
> 
> Also there is a resource handler overhaul in the pipeline (See https://github.com/hyprwm/hyprlock/pull/879).
> Meaning all the resource handling of the eyes and the password needs to be adapted.
> 
> Let me know if you have enough time to do that. If not we can collab a bit if you want.
> 
>

salfel commented:
I've now implemented the hiding using ctrl+h and done a little refactor to declutter the draw method, lmk if i should further move logic to separate function. 

Also want to repeat my question about the config option for the keybind for toggling.

PaideiaDilemma commented:
> Regarding the keybind, that seems fair. Maybe we could also add a config option to set the keybind for this along with other actions. Just a thought for later on

The thing is that right now hyprlock does not have any keybinding configurations anywhere.
So if we add that, in a different PR please.

> I've now implemented the hiding using ctrl+h and done a little refactor to declutter the draw method, lmk if i should further move logic to separate function.

Looks better to me than before, but I will look over it again, once you managed to pull in the resource manager changes.

Let me know if you need any help for that or if something I changed doesn't make sense to you :)

salfel commented:
i feel like it would then be better to just close this pr, as the whole code is pretty much just rendering logic, which would then need to be reimplemented anyways. I'd just create a pr after the merge. Feel like this makes more sense
